// pptx-automizer/dist/helper/modify-hyperlink-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HyperlinkElement {
    constructor(doc, relId, isInternal) {
        this.doc = doc;
        this.relId = relId;
        this.isInternal = isInternal;
    }
    createHlinkClick() {
        const hlinkClick = this.doc.createElement('a:hlinkClick');
        hlinkClick.setAttribute('r:id', this.relId);
        hlinkClick.setAttribute('xmlns:r', 'http://schemas.openxmlformats.org/officeDocument/2006/relationships');
        if (this.isInternal) {
            hlinkClick.setAttribute('action', 'ppaction://hlinksldjump');
            hlinkClick.setAttribute('xmlns:a', 'http://schemas.openxmlformats.org/drawingml/2006/main');
            hlinkClick.setAttribute('xmlns:p14', 'http://schemas.microsoft.com/office/powerpoint/2010/main');
        }
        return hlinkClick;
    }
    createTextRun(text) {
        const run = this.doc.createElement('a:r');
        const rPr = this.doc.createElement('a:rPr');
        const t = this.doc.createElement('a:t');
        rPr.appendChild(this.createHlinkClick());
        t.textContent = text;
        run.appendChild(rPr);
        run.appendChild(t);
        return run;
    }
}
exports.default = HyperlinkElement;
