// pptx-automizer/dist/pptx-automizer/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DxaToCm = exports.CmToDxa = exports.read = exports.modify = exports.LabelPosition = exports.ModifyImageHelper = exports.ModifyColorHelper = exports.ModifyTextHelper = exports.ModifyChartHelper = exports.ModifyTableHelper = exports.ModifyShapeHelper = exports.ModifyHelper = exports.XmlHelper = exports.Automizer = void 0;
const automizer_1 = __importDefault(require("./automizer"));
exports.Automizer = automizer_1.default;
const modify_helper_1 = __importStar(require("./helper/modify-helper"));
exports.ModifyHelper = modify_helper_1.default;
Object.defineProperty(exports, "CmToDxa", { enumerable: true, get: function () { return modify_helper_1.CmToDxa; } });
Object.defineProperty(exports, "DxaToCm", { enumerable: true, get: function () { return modify_helper_1.DxaToCm; } });
const modify_shape_helper_1 = __importDefault(require("./helper/modify-shape-helper"));
exports.ModifyShapeHelper = modify_shape_helper_1.default;
const modify_table_helper_1 = __importDefault(require("./helper/modify-table-helper"));
exports.ModifyTableHelper = modify_table_helper_1.default;
const modify_chart_helper_1 = __importDefault(require("./helper/modify-chart-helper"));
exports.ModifyChartHelper = modify_chart_helper_1.default;
const modify_hyperlink_helper_1 = __importDefault(require("./helper/modify-hyperlink-helper"));
const xml_helper_1 = require("./helper/xml-helper");
Object.defineProperty(exports, "XmlHelper", { enumerable: true, get: function () { return xml_helper_1.XmlHelper; } });
const modify_text_helper_1 = __importDefault(require("./helper/modify-text-helper"));
exports.ModifyTextHelper = modify_text_helper_1.default;
const modify_color_helper_1 = __importDefault(require("./helper/modify-color-helper"));
exports.ModifyColorHelper = modify_color_helper_1.default;
const modify_image_helper_1 = __importDefault(require("./helper/modify-image-helper"));
exports.ModifyImageHelper = modify_image_helper_1.default;
const chart_type_1 = require("./enums/chart-type");
Object.defineProperty(exports, "LabelPosition", { enumerable: true, get: function () { return chart_type_1.LabelPosition; } });
const dump = modify_helper_1.default.dump;
const dumpChart = modify_helper_1.default.dumpChart;
const setAttribute = modify_helper_1.default.setAttribute;
const setSolidFill = modify_shape_helper_1.default.setSolidFill;
const setText = modify_shape_helper_1.default.setText;
const setBulletList = modify_shape_helper_1.default.setBulletList;
const replaceText = modify_shape_helper_1.default.replaceText;
const setPosition = modify_shape_helper_1.default.setPosition;
const updatePosition = modify_shape_helper_1.default.updatePosition;
const rotateShape = modify_shape_helper_1.default.rotate;
const setTableData = modify_table_helper_1.default.setTableData;
const adjustHeight = modify_table_helper_1.default.adjustHeight;
const adjustWidth = modify_table_helper_1.default.adjustWidth;
const setTable = modify_table_helper_1.default.setTable;
const updateColumnWidth = modify_table_helper_1.default.updateColumnWidth;
const updateRowHeight = modify_table_helper_1.default.updateRowHeight;
const setRelationTarget = modify_image_helper_1.default.setRelationTarget;
const setDuotoneFill = modify_image_helper_1.default.setDuotoneFill;
const setChartData = modify_chart_helper_1.default.setChartData;
const setExtendedChartData = modify_chart_helper_1.default.setExtendedChartData;
const setChartVerticalLines = modify_chart_helper_1.default.setChartVerticalLines;
const setChartScatter = modify_chart_helper_1.default.setChartScatter;
const setChartBubbles = modify_chart_helper_1.default.setChartBubbles;
const setChartCombo = modify_chart_helper_1.default.setChartCombo;
const setAxisRange = modify_chart_helper_1.default.setAxisRange;
const setPlotArea = modify_chart_helper_1.default.setPlotArea;
const setLegendPosition = modify_chart_helper_1.default.setLegendPosition;
const removeChartLegend = modify_chart_helper_1.default.removeChartLegend;
const minimizeChartLegend = modify_chart_helper_1.default.minimizeChartLegend;
const setWaterFallColumnTotalToLast = modify_chart_helper_1.default.setWaterFallColumnTotalToLast;
const setChartTitle = modify_chart_helper_1.default.setChartTitle;
const setDataLabelAttributes = modify_chart_helper_1.default.setDataLabelAttributes;
const readWorkbookData = modify_chart_helper_1.default.readWorkbookData;
const readChartInfo = modify_chart_helper_1.default.readChartInfo;
const setHyperlinkTarget = modify_hyperlink_helper_1.default.setHyperlinkTarget;
const addHyperlink = modify_hyperlink_helper_1.default.addHyperlink;
const removeHyperlink = modify_hyperlink_helper_1.default.removeHyperlink;
exports.modify = {
    dump,
    dumpChart,
    setAttribute,
    setSolidFill,
    setText,
    setBulletList,
    replaceText,
    setPosition,
    updatePosition,
    rotateShape,
    setTableData,
    adjustHeight,
    adjustWidth,
    updateColumnWidth,
    updateRowHeight,
    setTable,
    setRelationTarget,
    setDuotoneFill,
    setChartData,
    setAxisRange,
    setExtendedChartData,
    setChartVerticalLines,
    setChartScatter,
    setChartCombo,
    setChartBubbles,
    setPlotArea,
    setLegendPosition,
    removeChartLegend,
    minimizeChartLegend,
    setWaterFallColumnTotalToLast,
    setChartTitle,
    setDataLabelAttributes,
    setHyperlinkTarget,
    addHyperlink,
    removeHyperlink,
};
exports.read = {
    readWorkbookData,
    readChartInfo,
};
exports.default = automizer_1.default;
