// pptx-automizer/dist/helper/archive/archive.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const xmldom_1 = require("@xmldom/xmldom");
class Archive {
    constructor(filename, params) {
        this.buffer = [];
        this.options = {
            type: 'nodebuffer',
        };
        this.filename = filename;
        this.params = params;
    }
    parseXml(xmlString) {
        const dom = new xmldom_1.DOMParser();
        return dom.parseFromString(xmlString, 'application/xml');
    }
    serializeXml(XmlDocument) {
        const s = new xmldom_1.XMLSerializer();
        return s.serializeToString(XmlDocument);
    }
    writeBuffer(archiveType) {
        return __awaiter(this, void 0, void 0, function* () {
            for (const buffered of this.buffer) {
                const serialized = this.serializeXml(buffered.content);
                yield archiveType.write(buffered.relativePath, serialized);
            }
        });
    }
    toBuffer(relativePath, content) {
        const existing = this.fromBuffer(relativePath);
        if (!existing) {
            this.buffer.push({
                relativePath,
                name: relativePath,
                content: content,
            });
        }
    }
    setOptions(params) {
        if (params.compression > 0) {
            this.options.compression = 'DEFLATE';
            this.options.compressionOptions = {
                level: params.compression,
            };
        }
    }
    fromBuffer(relativePath) {
        return this.buffer.find((file) => file.relativePath === relativePath);
    }
}
exports.default = Archive;
